E1744 Debouncing + improvements

vandalon opened the pull request:
I used the same method as with E1812. 
This requires a helper for all integrations.
Stop events are not debounced if to make sure a stop event is always processed. 
Added small HA-debounce to ensure automation restarts always happen.

Closes #20

EPMatt commented:
Hi @vandalon, thanks again! :)

Maybe it's better if we finalize the E1812 blueprint, then apply the same changes to the E1744, including improvements and refinements we are working on there. Meanwhile, we can leave this PR in draft.

What do you think about it?

vandalon commented:
Maybe we can push this one since it works for the one user that did had issues. If we push it we reach a bigger "user base" that can either confirms this works by not submitting any new issues or help us narrow new issues down :).
I'm using this controller actively now for light and audio on a daily basis and also for me it's stable :)

EPMatt commented:
Nice idea. :) 
I'll review your PR tomorrow, then we can merge it and make the feature available to other users as well.

Thanks again!

EPMatt commented on blueprints/controllers/ikea_e1744/ikea_e1744.yaml:
```diff
@@ -142,8 +157,6 @@ variables:
   click_triple: '{{ actions_mapping[integration_id]["click_triple"] }}'
-  # integrations which need to store the previous long press event to determine which button was released
-  integrations_with_prev_event_storage: [zha, zigbee2mqtt]
   # build data to send within a controller event
```
```suggestion
  # integrations which need to store the previous long press event to determine which button was released
  integrations_with_prev_event_storage: [zha, zigbee2mqtt]
  # build data to send within a controller event
```

We have to keep the `integrations_with_prev_event_storage` to identify whether the left or right rotation has stopped.

vandalon replied:
We do this anyway. But because we now use the helper for everything we don't need to define what integration uses it since they all use the helper and thus zha and zigbee2mqtt have this per definition.

EPMatt replied:
The `integrations_with_prev_event_storage` allows to avoid an additional condition in the `rotate_stop` actions for deCONZ users.

EPMatt commented on blueprints/controllers/ikea_e1744/ikea_e1744.yaml:
```diff
@@ -194,18 +217,21 @@ action:
+          {{ (states(helper_last_controller_event) | from_json).trigger_action }}
+        {%- endif -%}'
+  # save the last controller event, store the button event
+  - choose:
+    - conditions: '{{trigger_action != last_controller_event}}'
+      sequence:
+      - service: input_text.set_value
+        data:
+          entity_id: !input 'helper_last_controller_event'
+          value: '{"trigger_action":"{{trigger_action}}","last_triggered":"{{now()}}","last_event":"{{last_controller_event}}" }' 
```
```suggestion
  - service: input_text.set_value
    data:
      entity_id: !input 'helper_last_controller_event'
      value: '{"trigger_action":"{{trigger_action}}","last_triggered":"{{now()}}","last_event":"{{last_controller_event}}" }' 
```

Don't have a proper way to test this, but I guess that if we store the object only when the trigger is different we don't always get debouncing (if a user only clicks the remote, the `last_triggered` value will always be the one from the first press).

vandalon replied:
No, i need the last_triggered value to always be updated for debouncing and such.

vandalon replied:
Sorry, previous comment was nonsense. The reason I do it like this is because else the wait for trigger also causes a restart.

EPMatt replied:
Ok don't worry. We'll look into this later. :)

vandalon replied:
I remember now! since we ignore double events we also don't want to register them in the helper, because that influences the last_triggered time when it's actually ignored.

EPMatt commented on blueprints/controllers/ikea_e1744/ikea_e1744.yaml:
```diff
@@ -215,11 +241,9 @@ action:
           - repeat:
               sequence: !input action_rotate_left
-              until: '{{ not rotate_left_loop }}'
+              while: '{{ rotate_left_loop }}'
```
```suggestion
              until: '{{ not rotate_left_loop }}'
```

We have to keep using a post-test conditional loop to make sure the custom `rotate` action gets executed at least once, if the `rotate_left_loop` is set to `false`. In this case, the `rotate` action needs to be performed only once.

vandalon replied:
I don't think there is a difference in how it works. both check conditions every loop. But I don't like a negative (not) if it's not needed :)

EPMatt replied:
Actually there is:
> ### REPEAT UNTIL
> This form accepts a list of conditions that are evaluated **after** each time the sequence is run. Therefore **the sequence will always run at least once**. The sequence will be run until the condition(s) evaluate to true.



https://www.home-assistant.io/docs/scripts/#repeat-until

vandalon replied:
I see, sorry about that.

But in a while loop, you only want to execute if the event is still a rotate event right? So if we evaluate before the start of the loop and the event is no longer a rotate-event. it should stop there, no?

EPMatt commented on blueprints/controllers/ikea_e1744/ikea_e1744.yaml:
```diff
@@ -215,11 +241,9 @@ action:
           - '{{ trigger_action in rotate_left_stop }}'
-          # for integrations which need to store the last controller event, need to check the previous button event, stored in the provided input_text
-          - '{{ not integration_id in integrations_with_prev_event_storage or states(helper_last_controller_event) in rotate_left }}'
         sequence:
```
```suggestion
          # for integrations which need to store the last controller event, need to check the previous button event, stored in the provided input_text
          - '{{ not integration_id in integrations_with_prev_event_storage or last_controller_event in rotate_left }}'
        sequence:
```

Don't have a way to properly test this, but it should work to identify which stop action to execute.

vandalon replied:
I removed this because we need to make sure a condition is always matched, because otherwise it could happen a ahb_controller stop message is not send out causing continues loop.

EPMatt replied:
The issue is that if we receive a stop action, we'll always execute the `rotate_left_stop` action for ZHA or Zigbee2MQTT due to the order of choose steps. This is an erroneous behaviour, since there would not be a way to execute the `rotate_right_stop` action.

EPMatt commented on blueprints/controllers/ikea_e1744/ikea_e1744.yaml:
```diff
@@ -248,11 +265,9 @@ action:
           - repeat:
               sequence: !input action_rotate_right
-              until: '{{ not rotate_right_loop }}'
+              while: '{{ rotate_right_loop }}'
```
```suggestion
              until: '{{ not rotate_right_loop }}'
```
same as above

vandalon replied:
same as earlier :D

EPMatt commented on blueprints/controllers/ikea_e1744/ikea_e1744.yaml:
```diff
@@ -248,11 +265,9 @@ action:
           - '{{ trigger_action in rotate_right_stop }}'
-          # for integrations which need to store the last controller event, need to check the previous button event, stored in the provided input_text
-          - '{{ not integration_id in integrations_with_prev_event_storage or states(helper_last_controller_event) in rotate_right }}'
         sequence:
```
```suggestion
          # for integrations which need to store the last controller event, need to check the previous button event, stored in the provided input_text
          - '{{ not integration_id in integrations_with_prev_event_storage or last_controller_event in rotate_right }}'
        sequence:
```

same as above

vandalon replied:
Also same as earlier :)

EPMatt commented on blueprints/controllers/ikea_e1744/ikea_e1744.yaml:
```diff
@@ -281,7 +298,8 @@ action:
           - choose:
-            default: !input action_click_double
+              - conditions: []
+                sequence: !input 'action_click_double'
```
```suggestion
                sequence: !input action_click_double
```

EPMatt commented on blueprints/controllers/ikea_e1744/ikea_e1744.yaml:
```diff
@@ -271,7 +287,8 @@ action:
           - choose:
-            default: !input action_click_short
+              - conditions: []
+                sequence: !input 'action_click_short'
```
```suggestion
                sequence: !input action_click_short
```

EPMatt commented on blueprints/controllers/ikea_e1744/ikea_e1744.yaml:
```diff
@@ -261,7 +276,8 @@ action:
           - choose:
-            default: !input action_rotate_right_stop
+              - conditions: []
+                sequence: !input 'action_rotate_right_stop'
```
```suggestion
                sequence: !input action_rotate_right_stop
```

EPMatt commented on blueprints/controllers/ikea_e1744/ikea_e1744.yaml:
```diff
@@ -291,4 +309,5 @@ action:
           - choose:
-            default: !input action_click_triple
+              - conditions: []
+                sequence: !input 'action_click_triple'
```
```suggestion
                sequence: !input action_click_triple

```

EPMatt commented on blueprints/controllers/ikea_e1744/ikea_e1744.yaml:
```diff
@@ -228,17 +252,10 @@ action:
           - choose:
-            default: !input action_rotate_left_stop
+              - conditions: []
+                sequence: !input 'action_rotate_left_stop'
```
```suggestion
                sequence: !input action_rotate_left_stop
```

EPMatt commented on blueprints/controllers/ikea_e1744/ikea_e1744.yaml:
```diff
@@ -107,6 +121,7 @@ variables:
   helper_last_controller_event: !input helper_last_controller_event
   # integration id used to select items in the action mapping
   integration_id: '{{ integration | lower }}'
+  helper_debounce_delay: !input 'helper_debounce_delay'
```
```suggestion
  helper_debounce_delay: !input helper_debounce_delay
```

EPMatt reviewed (changes requested):
Just a few changes, but the core of your debouncing is great for me. :)

Thanks again!

vandalon commented:
What I will do is just test with your changes and then come back to you, is that ok? :)

EPMatt commented:
> What I will do is just test with your changes and then come back to you, is that ok? :)

Perfect, thank you @vandalon. :)

vandalon commented:
I am also considering a roteate_delay/debounce option that will make sure that a rotate_left followed by a rotate right within that time is ignored. (The controller is very sensitive and when turning one way, you still get some opposite turn events). What do you say?

EPMatt commented:
> I am also considering a roteate_delay/debounce option that will make sure that a rotate_left followed by a rotate right within that time is ignored. (The controller is very sensitive and when turning one way, you still get some opposite turn events). What do you say?
> 
> EDIT:
> 
> It might be enough to check that the last controller event is not the opposite turn without a stop in between.

Ok for the second option. Moreover, does the controller in your experience always fire a `rotate_stop` event when stopped?

vandalon commented:
The controller does seem to yes, also i don't debounce those messages just to make sure it really stops.

EPMatt commented:
> The controller does seem to yes, also i don't debounce those messages just to make sure it really stops.

Myabe it's better to debounce them,  making sure at least one gets executed. If this is not happening, users might experience duplicate runs for the `rotate_*_stop` actions.

vandalon commented:
Maybe we can debounce them in the stop_action part of the blueprint specifically. I will look into it. I don't want it to halt the whole automation because a stop action can come within debounce time, and if that one is not pushed trough, no stop action is performed.

vandalon commented:
Can you explain the failed check for me? I don't get it :)

vandalon commented:
> > The controller does seem to yes, also i don't debounce those messages just to make sure it really stops.
> 
> Myabe it's better to debounce them, making sure at least one gets executed. If this is not happening, users might experience duplicate runs for the `rotate_*_stop` actions.

Actually we already deal with this by ignoring all events that are the same as the one before :)

EPMatt commented:
> Can you explain the failed check for me? I don't get it :)

My bad in configuring CI for PRs. I'm currently working on it. :)

EPMatt commented:
> > > The controller does seem to yes, also i don't debounce those messages just to make sure it really stops.
> > 
> > 
> > Myabe it's better to debounce them, making sure at least one gets executed. If this is not happening, users might experience duplicate runs for the `rotate_*_stop` actions.
> 
> Actually we already deal with this by ignoring all events that are the same as the one before :)

Oh great, so we are debouncing them. :D

vandalon commented:
I see the weirdest behaviour. when printing the debounce_helper variable to the text helper I see that the value is set to 50 while running until the end, then it sets back to whatever i've configured. So when turning the dial, i see a constant value of 50, until stopping, then it's set to 500 in my case

vandalon commented:
So,

Here is my latest :)

I noticed that with the way we did debouncing I kept having issues with this controller because of it's sensitivity in some cases. For me in those cases a delay as debounce works better. But since this is very likely, very personal, I made it a choice.

select -100ms of debounce and you will get 100ms of delay before executing the last of the series of commands that was send within this 100ms.

select 100ms of debouncing and the automation will be started at once and other events that come in within this 100ms will be ignored. 

Other then that i've;
- Added indenting in the more complex conditions.
- Added checks to prevent the same turn event to get repeated (and thus restart the automation).
- Added check to prevent sudden direction change without a stop in between. This because of the sensitive nature of the controller that will easily send out an unintended direction change.

EPMatt commented:
@vandalon I succesfully implemented checks for Pull Requests. From now on GitHub Actions will check if the code submitted in the PR is valid and properly formatted. If this does not happen it will provide some feedback on which errors are present and how to solve them. :) 

For formatting issues you can run `npm run format` on the root directory of your local repository (run `npm install` first if you never did it).
I'm also fixing code-style issues on your other PRs (just for this time ;) ).

vandalon commented:
I'm working on a mac, don't even have nodes and rpm on there, but i Fixed the some stuff based on your checks, now all checks have passed :D

vandalon commented:
@EPMatt I was thinking about the whole `integrations_with_prev_event_storage` part.

I think we can do without, and in the process reduce the amount of code for this blueprint.
I will try and explain:

If we have a rotation left or right, we can assume that any stop following that is either left-stop or right-stop (depending if the rotation before was left or right). Even if deconz in this case is sending a right stop directly after a left turn, we should still treat it like a right stop since that means an event was lost somewhere. 

So basically I think we can just say if `trigger_event` == (`rotate_left_stop` OR `rotate_right_stop`) and `last_controller_event` == `rotate_right` then sent `ahb_controller_event` `rotate_right_stop`. and the same then for when rotating left.

This would mean we need less checking and we can also create one array called `rotate_stop` in which we add all stop events.

Then we just have 1 chose for this event and act on it. Sending out the ahb_controller_event for either left or right stop depending on `last_controller_event`.

I don't think this has any disadvantages and also make the automation a bit more robust since we stop on any stop event, and not a specific one. It might not happen ofter that the wrong stop event is received, but it still can happen.

vandalon commented:
Also, Do you think there is a better way for communication about this stuff then in the comment on GitHub? Maybe something like discord or so?

EPMatt commented:
> @EPMatt I was thinking about the whole `integrations_with_prev_event_storage` part.
> 
> I think we can do without, and in the process reduce the amount of code for this blueprint.
> I will try and explain:
> 
> If we have a rotation left or right, we can assume that any stop following that is either left-stop or right-stop (depending if the rotation before was left or right). Even if deconz in this case is sending a right stop directly after a left turn, we should still treat it like a right stop since that means an event was lost somewhere.
> 
> So basically I think we can just say if `trigger_event` == (`rotate_left_stop` OR `rotate_right_stop`) and `last_controller_event` == `rotate_right` then sent `ahb_controller_event` `rotate_right_stop`. and the same then for when rotating left.
> 
> This would mean we need less checking and we can also create one array called `rotate_stop` in which we add all stop events.
> 
> Then we just have 1 chose for this event and act on it. Sending out the ahb_controller_event for either left or right stop depending on `last_controller_event`.
> 
> I don't think this has any disadvantages and also make the automation a bit more robust since we stop on any stop event, and not a specific one. It might not happen ofter that the wrong stop event is received, but it still can happen.
> 
> I would like to know your thoughts :)

Hi @vandalon,

your solution looks great to me. I agree that this way the blueprint would be more robust and reliable, and we could remove the dependency with the text helper, if debouncing is not used at all. :)

By the way, were you able in your experience to spot a `rotate_left_stop` event when stopping a right rotation (or the opposite)?

I'll review your code more in detail later, but at first sight it looks good. 🚀 
Thank you so much for all the effort you are putting on this! Really appreciate. :)

EPMatt commented:
> Also, Do you think there is a better way for communication about this stuff than in the comment on GitHub? Maybe something like discord or so?

In my very little experience, GitHub is really great for working on new features and bugs, since you have the discussion strictly coupled with the code you are working on. You can easily link to code snippets and review them without having to move across different tools or services, which I find quite time-consuming and dispersive.

But that's just my opinion. I think it also depends from person to person.
Did you have a different experience with other open source projects? I'd be happy to hear your opinion on that. :)

vandalon commented:
> your solution looks great to me. I agree that this way the blueprint would be more robust and reliable, and we could remove the dependency with the text helper, if debouncing is not used at all. :)

No, since we then need the helper for determining a left or right stop :)
But I don't think that's an issue and it also make it's more generic in use.
The default debounce time is set to 0 btw.

> 
> By the way, were you able in your experience to spot a `rotate_left_stop` event when stopping a right rotation (or the opposite)?
Not specifically but i noticed missing events before, this could off course also been caused by the HA-bug you filed and is now fixed. Still it's a wireless protocol and things like that can happen.

vandalon commented:
> > your solution looks great to me. I agree that this way the blueprint would be more robust and reliable, and we could remove the dependency with the text helper, if debouncing is not used at all. :)
> 
> No, since we then need the helper for determining a left or right stop :)
> But I don't think that's an issue and it also make it's more generic in use.
> The default debounce time is set to 0 btw.
> 
> > By the way, were you able in your experience to spot a `rotate_left_stop` event when stopping a right rotation (or the opposite)?
> > Not specifically but i noticed missing events before, this could off course also been caused by the HA-bug you filed and is now fixed. Still it's a wireless protocol and things like that can happen.

vandalon commented:
> > Also, Do you think there is a better way for communication about this stuff than in the comment on GitHub? Maybe something like discord or so?
> 
> In my very little experience, GitHub is really great for working on new features and bugs, since you have the discussion strictly coupled with the code you are working on. You can easily link to code snippets and review them without having to move across different tools or services, which I find quite time-consuming and dispersive.
> 
> But that's just my opinion. I think it also depends from person to person.
> Did you have a different experience with other open source projects? I'd be happy to hear your opinion on that. :)


Sadly no, also i'm not a developer at all :) But i like direct ways of communication :D

vandalon commented:
Do you think we can make it so that we can use multiple controller devices for 1 controller automation? I have for example 2 dimmers for the living room in 2 places doing the same things.

EPMatt commented:
Hi @vandalon,

> No we can't, since we then need the helper for determining a left or right stop :)
> But I don't think that's an issue and it also make it's more generic in use.
> The default debounce time is set to 0 btw.

Oh right, sorry for my confusion. :) 

> Not specifically but i noticed missing events before, this could off course also been caused by the HA-bug you filed and is now fixed. Still it's a wireless protocol and things like that can happen.

Nice idea btw, there's nothing wrong in treating every stop event as a stop event for the rotation is currently occurrin. In the rare situation this issue might occur, it would ensure the controller is working as intended.

EPMatt commented:
> Sadly no, also i'm not a developer at all :) But i like direct ways of communication :D

Me too, sometimes I prefer a more direct way of communication especially when more discussion and interaction is required, but not having to jump between different social networks/tools while coding is the best in my opinion. :) 

Btw if you prefer I'm also available and active on the Home Assistant Community Forum. :)

EPMatt commented:
> Do you think we can make it so that we can use multiple controller devices for 1 controller automation? I have for example 2 dimmers for the living room in 2 places doing the same things.

Thank you for the suggestion. By the way I don't see a real benefit in implementing this proposal. I think the blueprint might become too complex for the average user, which often uses a single controller for a single device.  

Currently, adding a second controller only requires to create an additional automation (two if you are using a Hook as well), in my opinion a good tradeoff between blueprint simplicity and functionality.

sk115 commented:
> > Sadly no, also i'm not a developer at all :) But i like direct ways of communication :D
> 
> Me too, sometimes I prefer a more direct way of communication especially when more discussion and interaction is required, but not having to jump between different social networks/tools while coding is the best in my opinion. :)
> 
> Btw if you prefer I'm also available and active on the Home Assistant Community Forum. :)

Here just to add a +1 the benefit of these discussions taking place in public forums (ie: here or on the HA Community Forums) when at all possible, for the benefit of those who might go looking for answers on the same topic - as I, holding my symfonisk dial in need of debouncing, am doing right now 😄 

Also, thanks for all the work on this y'all - I'm new to blueprints, but I'm going to try copying over some of the edits here & seeing what happens. Thanks!

sk115 commented:
Tried it out with timeout set to 20ms, but ran into a JSONDecodeError pretty early on, in the trigger_action condition if I read this correctly- see attached tracelog.

This issue did not occur with my previous automation using the main release of the blueprint - maybe it's something to do with the check for the previous event / not falling back to the default correctly? 

[trace automation.controller_ikea_e1744_symfonisk_rotary_remote_2 2021-04-26T20_48_50.409310+00_00.json.txt](https://github.com/EPMatt/awesome-ha-blueprints/files/6380304/trace.automation.controller_ikea_e1744_symfonisk_rotary_remote_2.2021-04-26T20_48_50.409310%2B00_00.json.txt)

EPMatt commented:
Hi @sk115, thank you for joining the discussion! 

> Here just to add a +1 the benefit of these discussions taking place in public forums (ie: here or on the HA Community Forums) when at all possible, for the benefit of those who might go looking for answers on the same topic - as I, holding my symfonisk dial in need of debouncing, am doing right now

I totally agree with you, in my opinion open source is not only software, but related discussions, reports and questions as well. :)

>Tried it out with timeout set to 20ms, but ran into a JSONDecodeError pretty early on, in the trigger_action condition if I read this correctly- see attached tracelog.
>
> This issue did not occur with my previous automation using the main release of the blueprint - maybe it's something to do with the check for the previous event / not falling back to the default correctly?

This issue is relative to the `input_text` helper having a non-JSON content in it. The fix is in the #32 PR which implements debouncing for the E1812 blueprint, as well as other optimizations.

I'm now moving to apply the changes discussed there to the E1744 Controller blueprint. The goal here is to try to use the same debouncing for all the Controller blueprints, applying the one from #32 which seems to work pretty well.

I'll let you know as soon as we have a stable version of this blueprint.
Thank you for your patience and help!

EPMatt commented:
Hi @sk115,

I've applied changes discussed in #32 to this blueprint, and updated docs accordingly. 
Could you please test the E1744 blueprint on your system and check that everything is working as expected?

Thank you for your precious help!

andrmuel commented:
Hello @EPMatt & @vandalon 

I might be confused, but I'm having issues with this version. Home assistant refuses to use the blueprint with error `invalid template (TemplateSyntaxError: expected name or number) for dictionary value @ data['action'][1]['variables']['last_controller_event']. Got None`. 

Looks to me like backslash escaping is needed on line 240 inside regex_match("") (same as previous line for trigger_delta).

With that, home assistant can use the blueprint; however upon moving the controller, I run into the original issue – upon moving the controller, I get an endless stream of events and eventually 0 or maximum volume.

EPMatt commented:
Hi @andrmuel,
thank you so much for your tests. I fixed the issue with the `last_controller_event` regex.

> With that, home assistant can use the blueprint; however upon moving the controller, I run into the original issue – upon moving the controller, I get an endless stream of events and eventually 0 or maximum volume.

To enable debouncing, please add a value for the `helper_debounce_delay` input, since by default this is now set to 0, with the result of the feature being disabled.

By the way, I think we should update documentation to better state that debouncing is disabled by default, and provide guidance to users on how to enable it (or maybe we could enable it by default, but I don't think this might be a good idea since only some integrations require debouncing).

Could you please test again the blueprint on your instance and check that it's working as expected?
Thank you for your time and help. :)

andrmuel commented:
Hello @EPMatt 

Sorry if it was unclear; I meant with debounce delay I get the issue. Given that it should also be fixed by updates in HA itself, maybe it is a new bug; I didn't look into it.

The latest version from @vandalon works fine with the same settings (just replaced the file with [this](https://raw.githubusercontent.com/EPMatt/awesome-ha-blueprints/103dbba829061dd760be811f2ee1553b48b9d6e4/blueprints/controllers/ikea_e1744/ikea_e1744.yaml) and restarted).

EPMatt commented:
Hi @andrmuel,

thank you for your quick feedback. I've just identified the possible issue. It was a problem with the definition of `rotate_stop` events, which was probably causing them to not be recognized by the blueprint.

I've just pushed a fix, I hope it will run as expected this time. :) 
I'm sorry for the back and forth we're doing here but unfortunately I don't own an E1744 controller by myself, so I cannot test this on my system.

Thank you so much for your help. :)

andrmuel commented:
Hello @EPMatt 

I didn't realize you don't have this controller – all the more thanks for still supporting it!

I just tested your latest version, and it does work. The debounce is required for me (currently using 100 ms).

EPMatt commented:
Hi @andrmuel,

great! I'm glad to hear that the blueprint is working for you as expected. Huge thanks for your precious feedback. 🎉

>I didn't realize you don't have this controller – all the more thanks for still supporting it!

You're welcome, I'm happy to help even if I don't own all these controllers by myself. :)  

Thanks again!

EPMatt commented:
@vandalon since we have received positive feedback from @andrmuel I think we can now merge this PR.

If you want to propose other changes I'd be happy to hear your voice. I'll keep this PR open for two days for discussion, then if there are no comments, I'll merge it and make the feature publicly available. Many users are requesting this feature in the HA forum, so I'd prefer not to make them wait too long, since the blueprint is proven to be fully working. :)

Thank you for your great contribution to the project. :)

vandalon commented:
Hey Guys,

I had an insanely busy week, sorry about that. No I have a week of vacation so I should be able to come back on this the coming days.

EPMatt commented:
Hi @vandalon, 

don't worry about that, I'll keep this PR open so you can test and report when you're back.
Have a nice vacation! :)

vandalon commented:
Hi @EPMatt,

As far as I can see this seems to work ok. Maybe just commit it and see how it works for the rest of your users? :)

vandalon commented:
hmm, somehow I could create an endless loop with the latest code again. Where is my latest commit did not. I'll have to investigate why this happens again.

EPMatt commented:
Hi @vandalon, thank you so much for reporting. :)

> hmm, somehow I could create an endless loop with the latest code again. Where is my latest commit did not. I'll have to investigate why this happens again.

I guess a true endless loop won't be possible here, since we are limiting the maximum number of repetitions with the `rotate_right_max_loops` and `rotate_left_max_loops` inputs.

Moreover, it seems that the controller always fires one or more `rotate_stop` events after a rotation. Since the executed event is always the last one, I cannot see how it would be possible to achieve an endless loop.

EPMatt commented:
> As far as I can see this seems to work ok. Maybe just commit it and see how it works for the rest of your users? :)

@vandalon yeah, I think the feature is ready, except for this issue with the loop actions. We might take advantage of the feedback received by other users to check if anyone is able to reproduce the latest problem you are experiencing.

So, may I merge this PR in the `main` branch? :) 

Thank you!

vandalon commented:
Of course you may :) But i'm preparing a push for making sure stop messages are always let trough. So if you give me a few minutes, you can check that :)

vandalon commented:
Hmm i'm to tired now. You can commit if you want. 

I think we need to filter out double rotate messages in the primary conditions since a rotate should always be followed by a stop or an opposite rotate message. Also double stop messages can be filtered out since that should never happen any way. 

After that , I think we should exclude stop messages from the debounce delay to make sure a stop is always executed ( to prevent the endless loop). 

If i'm not mistaken i've already did that in my commit a few weeks back. So if you want you can take a look at that. Or I can do it somewhere this week :)

vandalon commented:
> Moreover, it seems that the controller always fires one or more `rotate_stop` events after a rotation. Since the executed event is always the last one, I cannot see how it would be possible to achieve an endless loop.

Yeah, what I think happens is that 1 stop message is received, sets the last_controller_event, then the next stop message is received, restarting the automation, then when coming to the execution of the stop event part it checks if the last event was a rotate, which its not, and thus the stop is not processed

EPMatt commented:
Hi @vandalon, sorry for my late response.

I agree that we should take a deeper look at this. I don't think we should exclude stop events from debouncing though; we have to debounce them too, to prevent duplicate action runs, but we should find a way to at least have one of them to be accepted as a valid event by the blueprint.
The issue might be that the controller fires the following or a similar sequence of events:
- `rotate_*`
- `rotate_*_stop`
- **`rotate_*`**

with the last `rotate_*` stopping the execution of the `rotate_*_stop` event due to debouncing, thus generating an invalid action run. But that's something we should investigate more in detail, and feedback from other users might be useful as well. So I think it's finally time to update the changelog and merge this PR into `main`. 🎉

We'll then review any incoming feedback to check whether other users are experiencing the same issue you described.  

Thank you so much for all the time and effort you've put into this feature! Definitely an amazing contribution to the project. ;)
